OpenDreamRuntime/Procs/Native/DreamProcNative.cs

```diff
@@ -8,6 +11,50 @@ static class DreamProcNative {
             DreamProcNativeRoot.DreamManager = IoCManager.Resolve<IDreamManager>();
 
             DreamObjectDefinition root = objectTree.GetObjectDefinition(DreamPath.Root);
+
+            Assembly asm = Assembly.GetExecutingAssembly();
+            
+            var q = from t in Assembly.GetExecutingAssembly().GetTypes()
+                    where t.Name == "DreamProcNativeRoot" && t.IsClass && t.Namespace == "OpenDreamRuntime.Procs.Native"
+                    select t;
+
+            foreach (var t in q)
+            {
+                foreach (var memberI in t.FindMembers(MemberTypes.Method, BindingFlags.Public | BindingFlags.Static, null, null)) {
+                    if (memberI.Name.StartsWith("NativeProc")) {
+                        var attr = memberI.GetCustomAttribute<DreamProcAttribute>();
+                        var args = memberI.GetCustomAttributes<DreamProcParameterAttribute>();
+
+                        var argumentNames = new List<string>();
+                        var defaultArgumentValues = new Dictionary<string, DreamValue>();
+                        foreach (var arg in args) {
+                            argumentNames.Add(arg.Name);
+                            if (arg.DefaultValue != null) {
+                                defaultArgumentValues.Add(arg.Name, new DreamValue(arg.DefaultValue));
+                            }
+                        }
+
+                        var methodI = t.GetMethod(memberI.Name);
+                        DreamProc proc = null;
+                        try {
+                            var func = methodI.CreateDelegate<NativeProc.HandlerFn>();
+                            proc = new NativeProc(attr?.Name, null, argumentNames, null, defaultArgumentValues, func);
+                        } catch { }
+
+                        try {
+                            var func2 = methodI.CreateDelegate<AsyncNativeProc.HandlerFn>();
+                            proc = new AsyncNativeProc(attr?.Name, null, argumentNames, null, defaultArgumentValues, func2.Invoke);
+                        } catch { }
+
+                        if (proc == null) {
+                            throw new System.Exception($"Cannot resolve signature for proc '{attr?.Name}' in DreamProcNativeRoot");
+                        }
```

Comment:
You can do `methodI.GetCustomAttribute<AsyncStateMachineAttribute>() != null` to check if the method is async
```suggestion
                        if (methodI.GetCustomAttribute<AsyncStateMachineAttribute>() != null) {
                            var func = methodI.CreateDelegate<AsyncNativeProc.HandlerFn>();
                            proc = new AsyncNativeProc(attr?.Name, null, argumentNames, null, defaultArgumentValues, func.Invoke);
                        } else {
                            var func = methodI.CreateDelegate<NativeProc.HandlerFn>();
                            proc = new NativeProc(attr?.Name, null, argumentNames, null, defaultArgumentValues, func);
                        }
```

Reply:
My intent was to match a signature with a constructor in a uniform way, if other signatures/constructors are added later or if more control over the exception that is thrown for failure was desirable (this code might be eventually moved out into a more reusable function). Going to commit this suggestion since these features are not too important right now and the original does throw out exception information for potential debugging.